browser/brave_cosmetic_resources_tab_helper.cc

```diff
@@ -0,0 +1,463 @@
+    content::WebContents* contents)
+    : WebContentsObserver(contents),
+    enabled_1st_party_cf_filtering_(false) {
+  if (BraveCosmeticResourcesTabHelper::observing_script_->empty()) {
+    *BraveCosmeticResourcesTabHelper::observing_script_ =
+        LoadDataResource(kCosmeticFiltersGenerated[0].value);
+  }
+}
+
+BraveCosmeticResourcesTabHelper::~BraveCosmeticResourcesTabHelper() {
+}
+
+void BraveCosmeticResourcesTabHelper::GetUrlCosmeticResourcesOnUI(
+    content::GlobalFrameRoutingId frame_id, const std::string& url,
+    bool do_non_scriplets, std::unique_ptr<base::ListValue> resources) {
+  if (!resources) {
+    return;
+  }
+  for (auto i = resources->GetList().begin();
+      i < resources->GetList().end(); i++) {
+    base::DictionaryValue* resources_dict;
+    if (!i->GetAsDictionary(&resources_dict)) {
+      continue;
+    }
+    std::string to_inject;
+    resources_dict->GetString("injected_script", &to_inject);
+    if (to_inject.length() > 1) {
+      auto* frame_host = content::RenderFrameHost::FromID(frame_id);
+      if (!frame_host)
+        return;
+      frame_host->ExecuteJavaScriptInIsolatedWorld(
+          base::UTF8ToUTF16(to_inject),
+          base::NullCallback(), ISOLATED_WORLD_ID_CHROME_INTERNAL);
+    }
+    // Working on css rules
+    if (!do_non_scriplets)
+      return;
+    CSSRulesRoutine(url, resources_dict, frame_id);
+  }
+}
+
+void BraveCosmeticResourcesTabHelper::CSSRulesRoutine(
+    const std::string& url_string, base::DictionaryValue* resources_dict,
+    content::GlobalFrameRoutingId frame_id) {
+  // Check are first party cosmetic filters enabled
+  const GURL url(url_string);
+  if (url.is_empty() || !url.is_valid() || IsVettedSearchEngine(url.host())) {
+    return;
+  }
+
+  Profile* profile = Profile::FromBrowserContext(
+    web_contents()->GetBrowserContext());
+  enabled_1st_party_cf_filtering_ =
+      brave_shields::IsFirstPartyCosmeticFilteringEnabled(
+          HostContentSettingsMapFactory::GetForProfile(profile), url);
+  base::ListValue* cf_exceptions_list;
+  if (resources_dict->GetList("exceptions", &cf_exceptions_list)) {
+    for (size_t i = 0; i < cf_exceptions_list->GetSize(); i++) {
+      exceptions_.push_back(cf_exceptions_list->GetList()[i].GetString());
+    }
+  }
+  base::ListValue* hide_selectors_list;
+  base::ListValue* force_hide_selectors_list = nullptr;
+  if (resources_dict->GetList("hide_selectors", &hide_selectors_list)) {
+    if (enabled_1st_party_cf_filtering_) {
+      force_hide_selectors_list = hide_selectors_list;
+    } else {
+      std::string cosmeticFilterConsiderNewSelectors_script =
+          "(function() {"
+            "let nextIndex = window.cosmeticStyleSheet.rules.length;"
+            "const selectors = [";
+      for (size_t i = 0; i < hide_selectors_list->GetSize(); i++) {
+        cosmeticFilterConsiderNewSelectors_script += "`" +
+            hide_selectors_list->GetList()[i].GetString() + "`";
```

Comment:
i think this can be turned into a JS injection using something like `\`"]; executeMaliciousCode(); dummyVariable=["` for the selector

a safer way to do this would be to add a function to the page content script which accepts a list of selector strings as an input parameter. the list could get passed to the extension via a custom `chrome.*` API

Reply:
Looks like [this](https://source.chromium.org/chromium/chromium/src/+/master:out/android-Debug/gen/components/autofill_assistant/browser/devtools/devtools/domains/dom.h;l=203?q=QuerySelectorAll&ss=chromium) might be a promising candidate?

Reply:
Is there possibly a Chromium API for generating a JS "escaped" string representation?

I guess another option would be just inserting the list as a JSON string like we currently do in the extension version.

Reply:
Why not just use the C++ APIs, that will use the same string values, and treat them in the same safe way, as the JS version?

Otherwise, i think the "make a JSON of it  and feed it to the JS APIs" approach sounds less scary than trying to make sure a string isn't JS

Reply:
addressed using a json array of strings

Reply:
@pes10k lmk if the new way looks good to you